docs/architecture/adr-001-signed-messages.md

```diff
@@ -0,0 +1,76 @@
+# ADR 001: Cosmos SDK Signed Messages
+
+## Changelog
+
+## Context
+
+Having the ability to sign messages off-chain has proven to be a fundamental aspect
+of nearly any blockchain. The notion of signing messages off-chain has many 
+added benefits such as saving on computational costs and reducing transaction
+throughput and overhead. Within the context of the Cosmos SDK, some of the major
+applications of signing such data includes, but is not limited to, providing a
+cryptographic secure and verifiable means of proving validator identity and
+possibly associating it with some other framework or organization. In addition,
+having the ability to sign Cosmos SDK messages with a Ledger device.
+
+A standardized protocol for hashing and signing messages that can be implemented
+by the Cosmos SDK and any third-party organization is needed the subscribes to
+the following:
+
+* A specification of structured data
+* A framework for encoding structured data
+* A cryptographic secure hashing and signing algorithm
+* A framework for supporting extensions and domain separation
+
+This record is only concerned with the rationale and the standardized implementation
+of Cosmos SDK signed messages. It does **not** concern itself with the concept of
+replay attacks as that will be left up to the higher-level application implementation.
+If you view signed messages in the means of authorizing some action or data, then 
+such an application would have to either treat this as idempotent or have mechanisms
+in place to reject known signed messages.
+
+TODO: Should we bake in replay protection into the protocol?
```

Comment:
I think not. Replay protection is only an issue with "executable" signed messages (i.e. transactions) - if future users need it for similar applications (a la 0x) it's easy to implement on top of this specification.

Reply:
What we *do* want is domain separation - messages signed with Cosmos should never be the same as transactions signed with Cosmos, or messages/transactions signed with the same key for any other application.

This means we need a prefix of some sort.

Reply:
Indeed, currently in the process of spec-ing this out.